File: dart/blockchain/lib/ledger/mempool.dart
import 'dart:async';

import 'package:blockchain/codecs.dart';
import 'package:blockchain/common/event_sourced_state.dart';
import 'package:blockchain/common/parent_child_tree.dart';
import 'package:blockchain/common/resource.dart';
import 'package:blockchain/consensus/local_chain.dart';
import 'package:blockchain/ledger/transaction_output_state.dart';
import 'package:blockchain_protobuf/models/core.pb.dart';
import 'package:ribs_core/ribs_core.dart';
import 'package:ribs_effect/ribs_effect.dart';

abstract class Mempool {
  Future<Set<Transaction>> read(BlockId currentHead);
  Future<void> add(Transaction transaction);
  Stream<MempoolChange> get changes;
}

class MempoolImpl extends Mempool {
  final Future<BlockBody> Function(BlockId) fetchBlockBody;
  final BlockSourcedState<Map<TransactionId, MempoolEntry>> eventSourcedState;
  final StreamController<MempoolChange> mempoolChangesController;
  final Duration expirationDuration;
  final LocalChain localChain;

  MempoolImpl(
    this.fetchBlockBody,
    this.eventSourcedState,
    this.mempoolChangesController,
    this.expirationDuration,
    this.localChain,
  );

  static Resource<MempoolImpl> make(
    FetchBlockBody fetchBlockBody,
    FetchTransaction fetchTransaction,
    ParentChildTree<BlockId> parentChildTree,
    BlockId currentEventId,
    Duration expirationDuration,
    LocalChain localChain,
  ) =>
      ResourceUtils.streamController(
              () => StreamController<MempoolChange>.broadcast())
          .flatMap((mempoolChangesController) {
        final state = <TransactionId, MempoolEntry>{};
        final _add = (Transaction transaction) => state[transaction.id] =
            MempoolEntry(transaction, DateTime.now().add(expirationDuration));
        final eventSourcedState =
            BlockSourcedState<Map<TransactionId, MempoolEntry>>(
          (state, blockId) async {
            final blockBody = await fetchBlockBody(blockId);
            blockBody.transactionIds.forEach(state.remove);
            return state;
          },
          (state, blockId) async {
            final blockBody = await fetchBlockBody(blockId);
            for (final transactionId in blockBody.transactionIds) {
              final transaction = await fetchTransaction(transactionId);
              _add(transaction);
            }
            return state;
          },
          parentChildTree,
          state,
          currentEventId,
          (p0) async => {},
        );
        return Resource.pure(MempoolImpl(
          fetchBlockBody,
          eventSourcedState,
          mempoolChangesController,
          expirationDuration,
          localChain,
        )).flatTap(
            (impl) => ResourceUtils.backgroundStream(impl._expirationStream));
      });

  Stream<TransactionId> get _expirationStream =>
      Stream.periodic(Duration(seconds: 10)).asyncMap((_) async {
        final now = DateTime.now();
        return eventSourcedState.useStateAt(await localChain.currentHead,
            (state) async {
          final toRemove = Map.fromEntries(
              state.entries.where((e) => e.value.addedAt.isBefore(now)));
          for (final entry in toRemove.entries) {
            state.remove(entry.key);
            mempoolChangesController
                .add(MempoolExpired(transaction: entry.value.transaction));
          }
          return toRemove.keys;
        });
      }).expand(identity);

  @override
  Future<void> add(Transaction transaction) async =>
      eventSourcedState.useStateAt(await localChain.currentHead, (state) async {
        state[transaction.id] =
            MempoolEntry(transaction, DateTime.now().add(expirationDuration));
        mempoolChangesController.add(MempoolAdded(transaction: transaction));
      });

  @override
  Future<Set<Transaction>> read(BlockId currentHead) =>
      eventSourcedState.useStateAt(currentHead,
          (state) async => state.values.map((e) => e.transaction).toSet());

  @override
  Stream<MempoolChange> get changes => mempoolChangesController.stream;
}

class MempoolEntry {
  final Transaction transaction;
  final DateTime addedAt;

  MempoolEntry(this.transaction, this.addedAt);
}

sealed class MempoolChange {}

class MempoolAdded extends MempoolChange {
  final Transaction transaction;

  MempoolAdded({required this.transaction});
}

class MempoolExpired extends MempoolChange {
  final Transaction transaction;

  MempoolExpired({required this.transaction});
}

File: scala/core/src/main/scala/blockchain/ledger/Mempool.scala
package blockchain.ledger

import blockchain.*
import blockchain.consensus.*
import blockchain.codecs.given
import blockchain.models.*
import cats.implicits.*
import cats.effect.{Async, Resource}
import fs2.concurrent.Topic
import org.typelevel.log4cats.Logger
import org.typelevel.log4cats.slf4j.Slf4jLogger

import java.time.Instant

trait Mempool[F[_]]:
  def read: F[List[Transaction]]
  def add(transaction: Transaction): F[Unit]
  def changes: fs2.Stream[F, MempoolChange]

object Mempool:
  class State(var entries: List[Entry])
  object State:
    def default: State = new State(Nil)
  case class Entry(transaction: Transaction, addedAt: Instant)
  type BSS[F[_]] = BlockSourcedState[F, State]

  def make[F[_]: Async](
      bss: BSS[F],
      localChain: LocalChain[F],
      fetchHeader: FetchHeader[F],
      bodyValidation: BodyValidation[F],
      clock: Clock[F],
      saveTransaction: Transaction => F[Unit]
  ): Resource[F, Mempool[F] with BlockPacker[F]] =
    Resource
      .make(Topic[F, MempoolChange])(_.close.void)
      .map(mempoolChangesTopic =>
        new MempoolImpl[F](
          bss,
          localChain,
          fetchHeader,
          bodyValidation,
          clock,
          mempoolChangesTopic,
          saveTransaction
        )
      )

  def makeBSS[F[_]: Async](
      initialState: F[State],
      initialBlockId: F[BlockId],
      blockIdTree: BlockIdTree[F],
      onBlockChanged: BlockId => F[Unit],
      fetchBody: FetchBody[F],
      fetchTransaction: FetchTransaction[F]
  ): Resource[F, BSS[F]] =
    Resource
      .pure(new MempoolBSS[F](fetchBody, fetchTransaction))
      .flatMap(bssImpl =>
        BlockSourcedState.make(
          initialState,
          initialBlockId,
          bssImpl.applyBlock,
          bssImpl.unapplyBlock,
          blockIdTree,
          onBlockChanged
        )
      )

sealed trait MempoolChange
object MempoolChange:
  case class Added(transaction: Transaction) extends MempoolChange
  case class Removed(transaction: Transaction) extends MempoolChange

class MempoolImpl[F[_]: Async](
    bss: Mempool.BSS[F],
    localChain: LocalChain[F],
    fetchHeader: FetchHeader[F],
    bodyValidation: BodyValidation[F],
    clock: Clock[F],
    mempoolChangesTopic: Topic[F, MempoolChange],
    saveTransaction: Transaction => F[Unit]
) extends Mempool[F]
    with BlockPacker[F]:

  private given logger: Logger[F] = Slf4jLogger.getLoggerFromName("Mempool")

  override def read: F[List[Transaction]] =
    localChain.currentHead
      .flatMap(
        bss
          .stateAt(_)
          .use(state =>
            retainValid(state.entries.map(_.transaction)).flatTap(updated =>
              Async[F].delay(state.entries = updated.flatMap(u => state.entries.find(_.transaction.id == u.id)))
            )
          )
      )

  override def add(transaction: Transaction): F[Unit] =
    localChain.currentHead
      .flatMap(headId =>
        bss
          .stateAt(headId)
          .use(s =>
            if (s.entries.exists(_.transaction.id == transaction.id))
              logger.info(show"Ignoring duplicate transaction id=${transaction.id}")
            else
              bodyValidation
                .validate(
                  FullBlockBody(s.entries.map(_.transaction) :+ transaction),
                  TransactionValidationContext(headId, 0, 0)
                )
                .leftMap(errors => new IllegalArgumentException(errors.head))
                .rethrowT >>
                saveTransaction(transaction) >>
                Async[F].delay(s.entries = s.entries :+ Mempool.Entry(transaction, Instant.now())) >>
                logger.info(show"Included transaction id=${transaction.id}")
          )
      )

  override def changes: fs2.Stream[F, MempoolChange] =
    mempoolChangesTopic.subscribeUnbounded

  override def streamed: fs2.Stream[F, FullBlockBody] =
    fs2.Stream
      .apply(())
      .merge(
        localChain.adoptions.void
          .merge(changes.void)
      )
      .evalMap(_ => read)
      .map(FullBlockBody(_))

  private def retainValid(
      transactions: List[Transaction]
  ): F[List[Transaction]] =
    for {
      headId <- localChain.currentHead
      head <- fetchHeader(headId)
      slot <- clock.globalSlot
      newTransactions <- transactions.foldLeftM(
        List.empty[Transaction]
      )((transactions, transaction) =>
        bodyValidation
          .validate(
            FullBlockBody(transactions :+ transaction),
            TransactionValidationContext(headId, head.height + 1, slot)
          )
          .fold(_ => transactions, _ => transactions :+ transaction)
      )
    } yield newTransactions

class MempoolBSS[F[_]: Async](fetchBody: FetchBody[F], fetchTransaction: FetchTransaction[F]):
  def applyBlock(state: Mempool.State, blockId: BlockId): F[Mempool.State] =
    fetchBody(blockId)
      .flatMap(body =>
        Async[F].delay(
          body.transactionIds.foreach(transactionId =>
            state.entries = state.entries.filterNot(_.transaction.id == transactionId)
          )
        )
      )
      .as(state)
  def unapplyBlock(state: Mempool.State, blockId: BlockId): F[Mempool.State] =
    fetchBody(blockId)
      .flatMap(_.transactionIds.traverse(fetchTransaction))
      .flatMap(transactions =>
        Async[F].delay(
          state.entries ++= transactions.filter(_.rewardParentBlockId.isEmpty).map(Mempool.Entry(_, Instant.now()))
        )
      )
      .as(state)
